File: assets/rt/flows/pipelineParallel/PipelineParallelFlow.java
package rt.flows.pipelineParallel;

import rt.FearlessError;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletableFuture;

/*
  The plan:
  1. Magic .step in actor flow-ops to be a non blocking dispatch to a subject
  2. Magic .stop in actor flows to dispatch a stop message (and maybe .join on the signal?)
  ...
  Step ??: Profit!
 */

public interface PipelineParallelFlow {
  final class WrappedSinkK implements base.flows._PipelineParallelSink_0 {
    public static WrappedSinkK $self = new WrappedSinkK();
    // TODO: sink id is not actually used, just here to make debugging easier during development
//    static long SINK_ID = 0;
    @Override public base.flows._Sink_1 $hash$imm(base.flows._Sink_1 original) {
//      return original;
      return new WrappedSink(original);
    }
  }
  final class WrappedSink implements base.flows._PipelineParallelSink_1 {
    final base.flows._Sink_1 original;
    Subject subject;
    public WrappedSink(base.flows._Sink_1 original) {
      this.original = original;

//      System.out.println("SPAWNING SUBJ: "+subjectId);
//      this.subject = spawn(msg -> {
//        System.out.println("SUBJ: "+subjectId+", Message received: "+msg);
//        original.$35$mut$(msg);
//      }, () -> {
//        System.out.println("Stop received (SUBJ "+subjectId+")");
//        original.stop$mut$();
//      });
      this.subject = new Subject(original);
    }

    @Override public base.Void_0 stopDown$mut() {
//      System.out.println("Stopping subj "+subjectId);
      try {
        subject.submit(Message.Stop.INSTANCE);
        subject.join();
      } catch (DeterministicFearlessError fe) {
        throw fe;
      } catch (Throwable exception) {
        var message = exception.getMessage();
        if (exception instanceof StackOverflowError) { message = "Stack overflowed"; }
        throw new RuntimeException(message, exception);
      }
      return base.Void_0.$self;
    }
    @Override public base.Void_0 $hash$mut(Object x$) {
//      System.out.println("SUBJ: "+subjectId+" GOT MSG: "+x$);
      this.subject.submit(x$);
      return base.Void_0.$self;
    }
    @Override public base.Void_0 pushError$mut(base.Info_0 info$) {
      this.subject.submit(new Message.Error(info$));
      return base.Void_0.$self;
    }
  }

  sealed interface Message {
    record Error(base.Info_0 info) implements Message {}
    final class Stop implements Message {
      static final Stop INSTANCE = new Stop();
      private Stop() {}
    }
  }

  final class Subject implements Runnable {
    private final base.flows._Sink_1 downstream;
    private final BlockingQueue<Object> buffer = new ArrayBlockingQueue<>(512);
    private final Thread worker;
    private boolean softClosed = false;
    private volatile Throwable exception = null;
    private volatile CompletableFuture<Void> onEmpty = null;

    public Subject(base.flows._Sink_1 downstream) {
      this.downstream = downstream;
      this.worker = Thread.ofVirtual().start(this);
      this.worker.setUncaughtExceptionHandler(((_, e) -> this.exception = e));
    }

    public void submit(Object msg) {
      var didSubmit = buffer.offer(msg);
      if (didSubmit) { return; }
      assert onEmpty == null || onEmpty.isDone() : "onEmpty should be clean if we're not awaiting it.";
      onEmpty = new CompletableFuture<>();
      onEmpty.join();
      submit(msg);
    }

    @Override public void run() {
      while (true) {
        Object msg;
        if (onEmpty != null && !onEmpty.isDone()) {
          msg = buffer.poll();
          if (msg == null) {
            onEmpty.complete(null);
            continue;
          }
        } else {
          try {
            msg = buffer.take();
          } catch (InterruptedException e) {
            throw new RuntimeException(e);
          }
        }

        if (msg == Message.Stop.INSTANCE) {
          downstream.stopDown$mut();
          break;
        }
        if (msg instanceof Message.Error info) {
          processError(info);
          continue;
        }
        processDataMsg(msg);
      }
    }

    public void join() {
      try {
        worker.join();
        downstream.stopDown$mut();
        if (this.exception != null) {
          switch (exception) {
            case RuntimeException re -> throw re;
            case Error error -> throw error;
            default -> throw new RuntimeException(exception);
          }
        }
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
    }

    private void processError(Message.Error info) {
      if (softClosed) {
        return;
      }
      softClosed = true;
      // If we're in a nested flow and the pushError throws, then we want to propagate that as a deterministic
      // exception (as opposed to a random error that we wrap in a RuntimeException to prevent it from being
      // caught by Try/1)
      try {
        downstream.pushError$mut(info.info());
      } catch (FearlessError err) {
        throw new DeterministicFearlessError(err.info);
      }
    }

    private void processDataMsg(Object data) {
      if (softClosed) {
        return;
      }
      try {
        downstream.$hash$mut(data);
      } catch (FearlessError err) {
        // Keep "accepting" new messages, but don't actually do anything with them because we're in an error state
        softClosed = true;

        // If we're in a nested flow and the pushError throws, then we want to propagate that as a deterministic
        // exception (as opposed to a random error that we wrap in a RuntimeException to prevent it from being
        // caught by Try/1)
        try {
          downstream.pushError$mut(err.info);
        } catch (FearlessError err1) {
          throw new DeterministicFearlessError(err1.info);
        }
      } catch (ArithmeticException err) {
        softClosed = true;
        downstream.pushError$mut(base.Infos_0.$self.msg$imm(rt.Str.fromJavaStr(err.getMessage())));
      }
    }
  }

  final class DeterministicFearlessError extends FearlessError {
    DeterministicFearlessError(base.Info_0 info) {
      super(info);
    }
  }
}

File: assets/rt-js/flows/pipelineParallel/PipelineParallelFlow.js
// rt/flows/pipelineParallel/PipelineParallelFlow.js
import { FearlessError } from "../../FearlessError.js";
import { base$$Void_0 } from "../../../base/Void_0.js";

/* -------------------- Messages -------------------- */
export const Message = {
  Error: class {
    constructor(info) { this.info = info; }
  },
  Stop: class Stop {
    static INSTANCE = new Stop();
  }
};

/* ---------------- Deterministic Error ---------------- */
export class DeterministicFearlessError extends FearlessError {
  constructor(info) { super(info); }
}

/* ---------------- WrappedSink ---------------- */
export class WrappedSink {
  constructor(original) {
    this.original = original;
    this.subject = new Subject(original);
  }

  stopDown$mut$0() {
    // console.log("[WS.stopDown] submit STOP → join");
    try {
      // Enqueue Stop synchronously and wait for this subject to drain.
      this.subject.submit(Message.Stop.INSTANCE);
      // console.log("[Subject.join] waiting");
      this.subject.join();
    } catch (err) {
      if (err instanceof DeterministicFearlessError) throw err;
      if (err instanceof FearlessError) throw err;
      const msg =
        err instanceof Error && /stack overflow/i.test(err.message)
          ? "Stack overflowed"
          : err.message;
      throw new Error(msg, { cause: err });
    }
    return base$$Void_0.$self;
  }

  $hash$mut$1(x$) {
    this.subject.submit(x$);
    return base$$Void_0.$self;
  }

  pushError$mut$1(info$) {
    this.subject.submit(new Message.Error(info$));
    return base$$Void_0.$self;
  }

  softClose() { this.subject.softClosed = true; }
}

/* ------------- Factory singleton ------------- */
export class WrappedSinkK {
  static $self = new WrappedSinkK();
  $hash$imm$1(original) { return new WrappedSink(original); }
}

/* ---------------- Subject ---------------- */
class BlockingQueue {
  constructor(capacity = 512) {
    this.capacity = capacity;
    this.buf = [];
    this.waitTakers = [];
    this.waitSpace = [];
  }

  offer(item) {
    // console.log("[BQ.offer]", item);
    if (this.buf.length < this.capacity) {
      if (this.waitTakers.length > 0) {
        // console.log("[BQ.offer] fulfilling taker");
        const take = this.waitTakers.shift();
        take(item);
      } else {
        this.buf.push(item);
      }
      return true;
    }
    return false;
  }

  async take() {
    if (this.buf.length > 0) return this.buf.shift();
    return new Promise(r => this.waitTakers.push(r));
  }

  poll() { return this.buf.length > 0 ? this.buf.shift() : null; }

  signalSpace() {
    if (this.waitSpace.length > 0) {
      const w = this.waitSpace.shift();
      w();
    }
  }
}

export class Subject {

  constructor(downstream) {
    this.downstream = downstream;
    this.buffer = new BlockingQueue(512);
    this.exception = null;
    this.softClosed = false;
    this.limitStopped = false; // ← NEW flag
    this._workerPromise = this._runLoop();
  }

  submit(msg) {
    if (this.softClosed && msg !== Message.Stop.INSTANCE) return;
    const ok = this.buffer.offer(msg);
    if (!ok) throw new Error("Subject buffer full");
  }

  async _runLoop() {
    try {
      while (true) {
        let msg = this.buffer.poll();
        if (msg == null) msg = await this.buffer.take();

        if (msg === Message.Stop.INSTANCE) {
          this.downstream.stopDown$mut$0(); // join flushes here
          break;
        }

        if (msg instanceof Message.Error) {
          this._processError(msg);
          continue;
        }

        this._processDataMsg(msg);
      }
    } catch (e) {
      this.exception = e;
    }
  }

  async join() {
    await this._workerPromise;
    if (this.exception) throw this.exception;
    return base$$Void_0.$self;
  }

  _processError(msg) {
    if (this.softClosed) return;
    this.softClosed = true;
    try {
      this.downstream.pushError$mut$1(msg.info);
    } catch (err) {
      if (err instanceof FearlessError)
        throw new DeterministicFearlessError(err.info);
      throw err;
    }
  }


  _processDataMsg(data) {
    console.log("[JOIN STAGE DATA]", data);
    if (this.softClosed) return;

    const d = this.downstream;
    try {
      d.$hash$mut$1(data);

      // detect limit operator and trigger downstream stop once
      if (!this.limitStopped && d.runner_m$ && d.remaining_m$) {
        const r = Number(d.remaining_m$?.get$mut$0?.() ?? 0);
        if (r <= 1) {
          this.limitStopped = true;
          console.log("[Subject] limit exhausted → stopping downstream ONCE");
          // Proactively tell the *downstream of the limit* to stop,
          // but keep THIS Subject running so it can still receive Stop.
          d.downstream_m$?.stopDown$mut$0?.();
          // DO NOT set softClosed here
        }
      }
    } catch (err) {
      this.softClosed = true;
      // existing error handling ...
    }
  }
}

/* ---------------- Export container ---------------- */
export const PipelineParallelFlow = {
  WrappedSinkK,
  WrappedSink,
  Message,
  Subject,
  DeterministicFearlessError
};
